test/replication-luatest/gh_8433_raft_is_candidate_test.lua

```diff
@@ -0,0 +1,123 @@
+        box_cfg = g.box_cfg,
+    }
+
+    g.proxy1 = proxy:new{
+        client_socket_path = server.build_listen_uri('server1_proxy'),
+        server_socket_path = server.build_listen_uri('server1'),
+    }
+
+    g.proxy2 = proxy:new{
+        client_socket_path = server.build_listen_uri('server2_proxy'),
+        server_socket_path = server.build_listen_uri('server2'),
+    }
+
+    luatest.assert(g.proxy1:start{force = true}, 'Proxy from 3 to 1 started')
+    luatest.assert(g.proxy2:start{force = true}, 'Proxy from 3 to 2 started')
+    g.box_cfg.replication[1] = server.build_listen_uri('server1_proxy')
+    g.box_cfg.replication[2] = server.build_listen_uri('server2_proxy')
+
+    g.server3 = g.replica_set:build_and_add_server{
+        alias = 'server3',
+        box_cfg = g.box_cfg,
+    }
+
+    g.replica_set:start()
+    g.replica_set:wait_for_fullmesh()
+end)
+
+g.after_all(function(g)
+    g.replica_set:drop()
+end)
+
+g.test_prevote_fail = function(g)
+    --
+    -- Test that applier failure doesn't start elections if the node
+    -- doesn't have enough quorum for doing that:
+    --     1. Fill leader_witness_map in order to not allow election_update_cb
+    --        start election on timeout.
+    --     2. Break applier to leader and wait until election pass
+    --        (4 * replication_timeout).
+    --     3. Break last applier and make sure, that election isn't started.
+    --
+    luatest.assert_equals(g.replica_set:get_leader(), g.server1)
+    local old_term = g.server1:get_election_term()
+    g.server3:exec(function()
+        box.cfg({election_mode = 'candidate'})
+    end)
+
+    -- We must be sure, that the server3 cannot start elections on timeout.
+    -- It's witness map should not be empty, when update_election_cb is executed
+    local replication_timeout = g.box_cfg.replication_timeout
+    local death_timeout =  replication_timeout * 4
+    luatest.helpers.retrying({timeout = death_timeout}, function()
+        if not g.server3:grep_log('leader is seen: true, state: follower') then
+            error("Witness map may still be empty")
+        end
+    end)
+
+    g.proxy1:pause()
+
+    -- Wait for the server3 to notice leader death and clear the
+    -- corresponding bit in leader_witness_map. The elections are not
+    -- supposed to start as the server2 says, that it can see the leader.
+    g.server3:exec(function(timeout)
+        local opt = {timeout = timeout * 3, delay = box.cfg.replication_timeout}
+        rawset(_G, 'wait_for_upstream_death_with_id', function(id)
+            luatest.helpers.retrying(opt, function()
+                local status = box.info.replication[id].upstream.status
+                if status ~= 'disconnected' and status ~= 'stopped' then
+                    error('Applier is still alive')
+                end
+            end)
+        end)
+
+        _G.wait_for_upstream_death_with_id(box.info.election.leader)
+        require('fiber').sleep(timeout)
```

Comment:
You're right, it's not needed. I was wrong about the need for a `sm_election_update_cb` call for clearing bit, which shows, whether the current node sees a leader or not. It's enough to wait for applier death, as it clears this bit too.

The problem is the fact, that the test is flaky again (it's not related to removing `fiber.sleep` as it can fail with or without it). Here're the logs:

```
2023-04-10 12:34:57.936 [57772] main/111/applier/unix/:/home/serpentian/Programming/tarantool/test/var/001_replication-luatest/server2_proxy.sock I> can't read row
2023-04-10 12:34:57.936 [57772] main/111/applier/unix/:/home/serpentian/Programming/tarantool/test/var/001_replication-luatest/server2_proxy.sock xrow.c:231 E> ER_INVALID_MSGPACK: Invalid MsgPack - packet header
2023-04-10 12:34:57.936 [57772] main/111/applier/unix/:/home/serpentian/Programming/tarantool/test/var/001_replication-luatest/server2_proxy.sock I> RAFT: fencing resumed
2023-04-10 12:34:57.936 [57772] main/111/applier/unix/:/home/serpentian/Programming/tarantool/test/var/001_replication-luatest/server2_proxy.sock I> LEADER is not seen: 3
2023-04-10 12:34:57.968 [57772] main I> SM_ELECTION_UPDATE_CB
2023-04-10 12:34:57.968 [57772] main I> RAFT: begin new election round
2023-04-10 12:34:57.968 [57772] main I> RAFT: bump term to 3, follow
2023-04-10 12:34:57.968 [57772] main I> RAFT: vote for 2, follow
```

<details>
<summary> Diff for lib/raft.c </summary>

```
diff --git a/src/lib/raft/raft.c b/src/lib/raft/raft.c
index dd28a182c..b8ab2a5bc 100644
--- a/src/lib/raft/raft.c
+++ b/src/lib/raft/raft.c
@@ -464,8 +464,10 @@ raft_notify_is_leader_seen(struct raft *raft, bool is_leader_seen,
 
        if (is_leader_seen)
                bit_set(&raft->leader_witness_map, source);
-       else if (bit_clear(&raft->leader_witness_map, source))
+       else if (bit_clear(&raft->leader_witness_map, source)) {
+               say_info("LEADER is not seen: %" PRIu32, source);
                raft_sm_election_update(raft);
+       }
 }
 
 /** Update raft state once leader is seen. */
@@ -967,6 +969,7 @@ static void
 raft_sm_election_update_cb(struct ev_loop *loop, struct ev_timer *timer,
                           int events)
 {
+       say_info("SM_ELECTION_UPDATE_CB");
        (void)events;
        struct raft *raft = timer->data;
        assert(timer == &raft->timer);
```

</details>

The applier to `server2` still fails sometimes before `g.proxy1:pause()` line and leader doesn't send heartbeat, but applier to the leader doesn't fail, and the server3 consider it to be alive and takes it into account in healthy quorum. So, server3 starts election. It's proper behavior, but causes test flakiness. I propose to increase `replication_timeout` even more, to `0.1`. Of course, it doesn't guarantee, that applier won't fail in such case, but it definitely happens rarely, and leader has more time to send heartbeat.

If you know better solution, help would be appreciated.

Reply:
Yep, I think there's nothing we can do about it. Setting `replication_timeout` to `0.1` seems fine. We use the same `replication_timeout` in other tests, so at least that's consistent with them.